actors/builtin/storage_market/multiset.go

```diff
@@ -0,0 +1,123 @@
+package storage_market
+
+import (
+	cid "github.com/ipfs/go-cid"
+	errors "github.com/pkg/errors"
+	cbg "github.com/whyrusleeping/cbor-gen"
+
+	"github.com/filecoin-project/specs-actors/actors/util/adt"
+)
+
+type Multiset struct {
+	mp *adt.Map
+	store adt.Store
+}
+
+// Interprets a store as a HAMT-based map of AMTs with root `r`.
+func AsMultiset(s adt.Store, r cid.Cid) *Multiset {
+	return &Multiset{mp: adt.AsMap(s, r), store: s}
+}
+
+// Creates a new map backed by an empty HAMT and flushes it to the store.
+func MakeEmptyMultiset(s adt.Store) (*Multiset, error) {
+	m, err := adt.MakeEmptyMap(s)
+	return &Multiset{m, s}, err
+}
+
+// Returns the root cid of the underlying HAMT.
+func (mm *Multiset) Root() cid.Cid {
+	return mm.mp.Root()
+}
+
+// Adds a value for a key.
+func (mm *Multiset) Add(key adt.Keyer, idx uint64) error {
```

Comment:
I think it would be better to name this `Set` or `Put` so the set-semantics are clear. It does a different thing to the ListMultimap (which would retain duplicates and insertion order).